### PR TITLE
`Tutorial groups`: Allow management of tutorial group sessions

### PR DESCRIPTION
**Pull request by stefanwaldhauser**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
The following user stories are implemented via this PR:

- As an instructor I want to create,edit,delete the sessions of the tutorial groups of the course.
- As a tutor I want to create,edit,delete the sessions of the tutorial groups for which I am assigned.

- As an instructor I want to cancel / activate tutorial group sessions.
- As a tutor I want to cancel / activate tutorial group sessions for which I am assigned

- As a instructor I want to mark certain days as "tutorial group free days" where no sessions are happening because of a national holiday for example

- As a student I want to see the sessions of my registered tutorial groups but also of other tutorial groups in the course.

### Description

The most important new entity is **TutorialGroupSession** which represents one session of a particular tutorial group. For example Monday 24.10 11:00-12:00 in Room 303. A session is either _Active_ or _Cancelled._ The user who cancels a session can give an optional reason for the cancellation.

A session is either individual or follows the typical schedule of a tutorial group, meaning it is a recurring session. The typical schedule (every week on Wednesday from 11:00 to 12:00 in Room 302) of a tutorial group is represented by the new entity **TutorialGroupSchedule**. It is not mandatory for a tutorial group to have a fixed schedule. Every session can be instead created individually if desired. 

The schedule is valid for a certain period, for example 18.10.2021 – 11.02.2022. This information is used to generate the individual session entities of the schedule. So you can think of the schedule as a blueprint for creating the weekly, biweekly sessions of a tutorial group and so on.

Most of the time the tutorial groups of a course occur in the same period, so to simplify schedule creation, the entity **TutorialGroupsConfiguration** was created, which bundles all course wide configuration for tutorial groups. At the moment it only contains the common tutorial group period which will be used as a default value for the valid period of a schedule.

The instructor can also specify days where no tutorial group sessions occur. For example because of a national holiday. This information is saved in the the entity **TutorialFreePeriod**. Currently in the client we only support creating free days, so start represents beginning of the day and end the end of the day. 

An interesting discussion is how to handle scheduling changes or conflicts. I have made the following decisions regarding this topic:
- If the user changes the location and/or the datetime of a recurring session it will be converted to an individual session by cutting the connection to the schedule of a tutorial group. 
- If the user changes the schedule of a tutorial group, individual sessions will be untouched but all recurring sessions will be deleted and recreated according to the new schedule (for example now on monday instead of wednesday)
- Overlapping sessions of one tutorial group are forbidden.
- If the instructor adds a tutorial free day all sessions on that day will be automatically cancelled. If the instructor removes a tutorial free day, all cancelled sessions on that will be automatically activated.
- If a user created a session that overlaps with a tutorial free day, it will be created but automatically cancelled.

#### _Let us talk time zones for a moment_ 

At the moment we do not store time zone information in Artemis. This means, for example, that we always send dates to the server in the time zone of the client, i.e. the web browser, and the server then converts the value to UTC for storage. On the other hand, we always display data from the server in the time zone of the client, i.e. the browser.

**This is not necessarily wrong but leads to some usability problems, especially when it comes to scheduling issues:** 

"I am the leader of a course in Munich at TUM (timezone Europe/Berlin) and I am currently on a conference in Ecuador (timezone Ecuador Time). I would like to use the evening to plan the tutorial groups for the next semester. A tutorial group should always take place on Wednesdays from 10:00 to 12:00. However, since we do not store time zone information, I need to convert this to the Ecuador time zone. 10:00 in Munich is 3:00 in Ecuador."

It is the same for all other planning questions. Artemis shows the instructor from the example above all dates in Ecuador time and he/she has to rethink and recalculate all the time. 

**That's why I assigned the new attribute "time zone" to the course. At the moment this information is only used for exercise groups.**

- All dates in the **management view** of the tutorial groups are **displayed in the time zone of the course** no matter where the user is located, i.e. in the example from above in Europe/Berlin to facilitate planning.

- All dates in the **student view** are displayed in the time zone of the client, so that the student visiting his/her parents at home in Turkey knows at what time his tutorial group takes place.

An interesting question is how we should handle the case where a instructors changes the time zone of a course (should happen very very rarely). At the moment I decided to delete **all **sessions and tutorial free periods** and then recreate the recurring sessions according to their schedule in the new time zone (all individual sessions stay deleted and need to be recreated by the user). 

image

### Code Review
As this PR is too big for a thorough code review, **please take the time to look at files that touch existing parts of Artemis. The main goal should be to prevention of any bugs into existing parts of Artemis.

### Steps for Testing
Prerequisites:
- Confirm that the user stories from above work. (Creating a schedule, creating sessions, cancelling sessions, activating sessions and so on) (See also screenshots)

#### Code Review
- [x] Review 1
- [ ] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
Server:
| Class/File                          | Line | Branch |
|-------------------------------------|------|--------|
| TutorialGroupService                | 92   | 78     |
| TutorialGroupScheduleService        |   80 |     79 |
| TutorialGroupFreePeriodService      | 100  |     50 |
| TutorialGroupsConfigurationService  | 100  |    100 |
|                                     |      |        |
| TutorialGroupResource               | 85   |     62 |
| TutorialGroupFreePeriodResource     | 90   |     59 |
| TutorialGroupSessionResource        | 93   | 65     |
| TutorialGroupsConfigurationResource | 83   | 55     |
| TutorialGroupDateUtil               | 62   | 50     |

Client: 
| Class/File                                                                                                                                          | Line   | Branch |
|-----------------------------------------------------------------------------------------------------------------------------------------------------|--------|--------|
| app/overview/course-tutorial-groups/course-tutorial-groups-registered                                                                               |   100% |   100% |
| app/overview/course-tutorial-groups/course-tutorial-groups-overview                                                                                 |   100% |    75% |
| app/overview/course-tutorial-groups/course-tutorial-group-card                                                                                      |   100% |    75% |
| app/overview/course-tutorial-groups                                                                                                                 | 82.95% | 78.33% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-groups/tutorial-groups-management/tutorial-groups-import-dialog                      | 97.27% | 76.21% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-groups/tutorial-groups-management/tutorial-groups-import-button                      |   100% |    75% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-groups/tutorial-groups-management/tutorial-groups-course-information                 |   100% |    75% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-groups/tutorial-groups-management/tutorial-group-row-buttons                         | 95.23% |    75% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-groups/tutorial-groups-management                                                    | 91.11% |    70% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-groups/detail                                                                        |   100% | 76.47% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-groups/crud/tutorial-group-form/schedule-form                                        | 92.68% |    75% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-groups/crud/tutorial-group-form                                                      | 88.07% | 69.76% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-groups/crud/edit-tutorial-group                                                      | 88.88% | 65.51% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-groups/crud/create-tutorial-group                                                    | 97.67% | 80.95% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-groups-configuration/crud/tutorial-groups-configuration-form                         |   100% |  87.5% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-groups-configuration/crud/edit-tutorial-groups-configuration                         | 94.28% |    76% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-groups-configuration/crud/create-tutorial-groups-configuration                       | 94.44% | 75.86% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-groups-checklist                                                                     | 93.93% |    75% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-group-sessions/tutorial-group-sessions-management/tutorial-group-session-row-buttons | 90.62% | 66.66% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-group-sessions/tutorial-group-sessions-management/cancellation-modal                 | 84.84% | 76.47% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-group-sessions/tutorial-group-sessions-management                                    | 94.44% | 78.94% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-group-sessions/crud/tutorial-group-session-form                                      |   100% |  87.5% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-group-sessions/crud/edit-tutorial-group-session                                      | 95.12% | 75.67% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-group-sessions/crud/create-tutorial-group-session                                    | 94.59% |    75% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-free-periods/tutorial-free-periods-management/tutorial-group-free-period-row-buttons | 95.23% |    75% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-free-periods/tutorial-free-periods-management                                        | 97.14% | 77.27% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-free-periods/crud/tutorial-free-period-form                                          |   100% |  87.5% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-free-periods/crud/edit-tutorial-group-free-period                                    | 94.87% | 75.86% |
| app/course/tutorial-groups/tutorial-groups-management/tutorial-free-periods/crud/create-tutorial-group-free-period                                  |    96% |    75% |
| app/course/tutorial-groups/tutorial-groups-management/registered-students                                                                           | 83.05% | 68.57% |
| app/course/tutorial-groups/shared/tutorial-groups-table                                                                                             |   100% |    75% |
| app/course/tutorial-groups/shared/tutorial-group-sessions-table/tutorial-group-session-row                                                          |   100% |   100% |
| app/course/tutorial-groups/shared/tutorial-group-sessions-table                                                                                     |   100% | 84.21% |
| app/course/tutorial-groups/shared/tutorial-group-free-days-overview                                                                                 | 94.73% | 77.77% |
| app/course/tutorial-groups/shared/tutorial-group-detail                                                                                             | 95.23% | 83.33% |
| app/course/tutorial-groups/shared                                                                                                                   | 96.42% | 91.66% |
| app/course/tutorial-groups/services                                                                                                                 | 84.37% | 57.77% |

### Screenshots

_Instructor_

**Checklist when opening the tutorial group management on a course**

Before the tutorial group feature can be used, a time zone has to be added to the course and a course wide configuration created:

image

New Field for time zone in course update:
image

**NOTE: Here we select the time zone Europe/London while I am in Europe/berlin**

Create Tutorial Group Configuration
image

Once the course has a time zone and a course wide configuration was created, the instructor can start creating tutorial groups.
image

image

The instructor can modify individual sessions of all tutorial groups. **Note that all dates are displayed in the course time zone Europe / London for the management view**

image

image

image

As soon as one of the recurring sessions is modified (here time change) it will be converted to a individual session

image

image

In the management view instructors can also create course wide holidays. All sessions that fall on this date will be automatically cancelled

image

image

**Currently from the client side only single days are possible, but this will be extended such that you can select whole periods like christmas as tutorial free**

_Tutor_

The tutor has full power over the sessions and registrations of his/her tutorial group. 

image

_Students_

Lets quickly check out how the sessions look from a students perspective. **Important: Now every date/time is shown in the local time of the user which is in my case Europe/Berlin instead of Europe/London**

image

Clicking on the card brings them to detail page where again they see all sessions in their own time zone

image

### CODE REVIEW

**Review by JohannesStoehr — COMMENTED** [review]
Skimmed through the code and left some comments

File: src/main/java/de/tum/in/www1/artemis/domain/Course.java [hunk 1]
```diff
@@ -167,6 +168,12 @@ public class Course extends DomainObject {
     @JsonView(QuizView.Before.class)
     private Integer accuracyOfScores;
 
+    /**
+     * Note: Currently just used in the scope of the tutorial groups feature
+     */
+    @Column(name = "time_zone")
+    private String timeZone;
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/Course.java:175]: That feels kind of wrong in the time of cloud hosting tbh.

File: src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroupSchedule.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    private TutorialGroup tutorialGroup;
+
+    @Column(name = "day_of_week")
+    private Integer dayOfWeek;
+
+    /**
+     * Note: String to prevent Hibernate from converting it to UTC
+     */
+    @Column(name = "start_time")
+    private String startTime;
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroupSchedule.java:38]: There is a good discussion here: https://stackoverflow.com/questions/19626177/how-to-store-repeating-dates-keeping-in-mind-daylight-savings-time

The important part:

```
UTC is not necessarily the best way to store all date and time values. It works particularly well for past events, or for future absolute events, but it doesn't work so great for future local events - especially future recurring events.

I wrote about this [on another answer](https://stackoverflow.com/a/19170823/634824) recently, and is one of the few exception cases where local time makes more sense than UTC. The main argument is the "alarm clock problem". If you set your alarm clock by UTC, you'll be waking up an hour early or late on the day of the DST transitions. This is why most folks set their alarm clocks by local time.

```

[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroupSchedule.java:38]: **If you set your alarm clock by UTC, you'll be waking up an hour early or late on the day of the DST transitions. This is why most folks set their alarm clocks by local time.**

File: src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupDateUtil.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+     *
+     * @param timeString the string to check
+     * @return true if the string follows the ISO 8601 format for time
+     */
+    public static boolean isIso8601TimeString(String timeString) {
+        try {
+            LocalTime.parse(timeString);
+            return true;
+        }
+        catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupDateUtil.java]: Done in 700f64dbc3947beb5c0ae367678e6ef0c4c42e03

File: src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupDateUtil.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+     *
+     * @param dateString the string to check
+     * @return true if the string follows the ISO 8601 format for date
+     */
+    public static boolean isIso8601DateString(String dateString) {
+        try {
+            LocalDate.parse(dateString);
+            return true;
+        }
+        catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/web/rest/tutorialgroups/TutorialGroupDateUtil.java]: Done in 700f64dbc3947beb5c0ae367678e6ef0c4c42e03

**Review by tobias-lippert — COMMENTED** [review]
Tested during testing session on ts1.
Functionality worked as expected. Discussed some improvements regarding the UI as well as additional functionality that could be implemented.

File: src/main/java/de/tum/in/www1/artemis/service/tutorialgroups/TutorialGroupScheduleService.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        for (var individualSession : individualSessions) {
+            var overlappingSession = tutorialGroupSessionRepository.findOverlappingIndividualSessionsInSameTutorialGroup(tutorialGroup, individualSession.getStart(),
+                    individualSession.getEnd());
+            if (!overlappingSession.isEmpty()) {
+                overlappingIndividualSessions.addAll(overlappingSession);
+            }
+        }
+        if (!overlappingIndividualSessions.isEmpty()) {
+            throw new ScheduleOverlapsWithSessionException(overlappingIndividualSessions, ZoneId.of(tutorialGroupsConfiguration.getCourse().getTimeZone()));
+        }
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/tutorialgroups/TutorialGroupScheduleService.java:63]: Consider moving this into a method.

File: src/main/java/de/tum/in/www1/artemis/domain/Course.java [hunk 6]
```diff
@@ -204,6 +211,11 @@ public class Course extends DomainObject {
     @JsonIgnoreProperties("consecutiveCourses")
     private Set<LearningGoal> prerequisites = new HashSet<>();
 
+    @OneToOne(cascade = CascadeType.REMOVE, orphanRemoval = true, fetch = FetchType.LAZY)
+    @JoinColumn(name = "tutorial_groups_configuration_id")
+    @JsonIgnoreProperties(value = "course")
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/Course.java]: Done in 5f60dcc6f88f8ce79bd8a635caa26e7f7cdf1479

File: src/main/java/de/tum/in/www1/artemis/domain/enumeration/TutorialGroupSessionStatus.java [hunk 7]
```diff
@@ -0,0 +1,5 @@
+package de.tum.in.www1.artemis.domain.enumeration;
+
+public enum TutorialGroupSessionStatus {
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/enumeration/TutorialGroupSessionStatus.java:6]: Done in 2daffd536722afa29c3b8b9116947753cc366c9f

File: src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroupSchedule.java [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+     * Currently represents weekly recurrence, so 1 means every week, 2 means every other week, etc.
+     */
+    @Column(name = "repetition_frequency")
+    private Integer repetitionFrequency;
+
+    /**
+     * Note: String to prevent Hibernate from converting it to UTC
+     */
+    @Column(name = "valid_from_inclusive")
+    private String validFromInclusive;
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroupSchedule.java:82]: I added an explaining javadoc at: 29f5bfecbb8b4c0af365fc527cf98cf999e8e751

[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroupSchedule.java:82]: IMO no property can be removed. I checked what other calendar apps save for recurrence patterns and it is basically the same

File: src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroupSession.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+        }
+    }
+
+    /**
+     * Removes circular references for JSON serialization.
+     *
+     * @param tutorialGroupSession the tutorial group session to remove circular references for
+     * @return the tutorial group session without circular references
+     */
+    public static TutorialGroupSession preventCircularJsonConversion(TutorialGroupSession tutorialGroupSession) {
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroupSession.java:133]: I had the problem that @JsonIgnoreProperties just odes none level basically. But the connection Schedule -> TutorialGroup -> Session was still causing the conversion problem. I could try again with @JsonIgnorePropeties but this seems fine and does not cause any problems. So we can keep it imo until we find a cleaner solution

File: src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroupSession.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    @Column(name = "end")
+    private ZonedDateTime end;
+
+    @Enumerated(EnumType.STRING)
+    @Column(name = "status")
+    private TutorialGroupSessionStatus status;
+
+    @Column(name = "status_explanation")
+    @Size(min = 1, max = 256)
+    private String statusExplanation;
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/domain/tutorialgroups/TutorialGroupSession.java:51]: I added a explaining javadoc in cadc27da266e3de8f2fbf2a1cc7c31d343b652ae. Hopefully it is now more clear

File: src/main/java/de/tum/in/www1/artemis/service/tutorialgroups/TutorialGroupScheduleService.java (hunk 5 above)
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/service/tutorialgroups/TutorialGroupScheduleService.java]: Good idea! Done in 0a94757fc35450f1ba895a7ac0260d3bd3d8515b

File: src/main/java/de/tum/in/www1/artemis/service/tutorialgroups/TutorialGroupFreePeriodService.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+     * @param course                  the course in which the free period is defined
+     * @param tutorialGroupFreePeriod the free period
+     */
+    public void cancelActiveOverlappingSessions(Course course, TutorialGroupFreePeriod tutorialGroupFreePeriod) {
+        var overlappingSessions = tutorialGroupSessionRepository.findAllActiveBetween(course, tutorialGroupFreePeriod.getStart(), tutorialGroupFreePeriod.getEnd());
+        overlappingSessions.forEach(session -> {
+            session.setStatus(TutorialGroupSessionStatus.CANCELLED);
+            if (tutorialGroupFreePeriod.getReason() != null) {
+                session.setStatusExplanation(tutorialGroupFreePeriod.getReason());
+            }
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/service/tutorialgroups/TutorialGroupFreePeriodService.java]: Very good idea! I made this change in: 8678f55a3c917ceecba88091864ed1af58dd94fb

File: src/main/java/de/tum/in/www1/artemis/service/tutorialgroups/TutorialGroupFreePeriodService.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+     * @param course                  the course in which the free period is defined
+     * @param tutorialGroupFreePeriod the free period
+     */
+    public void activateCancelledOverlappingSessions(Course course, TutorialGroupFreePeriod tutorialGroupFreePeriod) {
+        var overlappingSessions = tutorialGroupSessionRepository.findAllCancelledBetween(course, tutorialGroupFreePeriod.getStart(), tutorialGroupFreePeriod.getEnd());
+        overlappingSessions.forEach(session -> {
+            session.setStatus(TutorialGroupSessionStatus.ACTIVE);
+            session.setStatusExplanation(null);
+        });
+        tutorialGroupSessionRepository.saveAll(overlappingSessions);
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/service/tutorialgroups/TutorialGroupFreePeriodService.java]: I changed this in 
image

I named it explanationStatus to keep the interpretation open if we introduce more values for status in the future. But I added an explaining javadoc that should clear things up.

**Pull request merged**